cmd/kubelet/app/server.go

```diff
@@ -930,21 +929,15 @@ func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, nod
 	}
 
 	kubeClientConfigOverrides(s, clientConfig)
-	closeAllConns, err := updateDialer(clientConfig)
-	if err != nil {
-		return nil, nil, err
-	}
-	return clientConfig, closeAllConns, nil
-}
 
-// updateDialer instruments a restconfig with a dial. the returned function allows forcefully closing all active connections.
-func updateDialer(clientConfig *restclient.Config) (func(), error) {
-	if clientConfig.Transport != nil || clientConfig.Dial != nil {
-		return nil, fmt.Errorf("there is already a transport or dialer configured")
+	type closeIdler interface {
+		CloseIdleConnections()
+	}
+	tr, ok := clientConfig.Transport.(closeIdler)
+	if !ok {
+		return nil, nil, fmt.Errorf("transport doesn't implement closeIdler: %w", err)
 	}
-	d := connrotation.NewDialer((&net.Dialer{Timeout: 30 * time.Second, KeepAlive: 30 * time.Second}).DialContext)
-	clientConfig.Dial = d.DialContext
-	return d.CloseAll, nil
+	return clientConfig, tr.CloseIdleConnections, nil
```

Comment:
I think there's still a hole if we only close idle connections

1. shared client is used by multiple goroutines
2. 2 connections are opened
3. both connections break (but don't close)
4. heartbeat fails on conn 1, while another goroutine is stuck waiting for its timeout to fire on conn 2
5. the heartbeat failure triggers calling CloseIdleConnections, which closes conn 1 but not conn 2
6. the heartbeat is reattempted, establishes a new working conn 3, and succeeds
7. the request over conn 2 finally times out, the request fails, and conn 2 is returned to the idle pool

That leaves us in a state where the heartbeat can be succeeding but we left broken connections in the idle pool

Reply:
is this about http2,  http1.1 or both? is the http2 hearbeat the one you mention?
in http2 you can not have 2 connections for the same destination.

Reply:
http/1.1

Reply:
the heartbeat is the node heartbeat goroutine... when that fails, it calls the "close all connections" function returned here

Reply:
> the request over conn 2 finally times out, the request fails, and conn 2 is returned to the idle pool

yeah, I can reproduce this behavior

Reply:
@liggitt http1.1 connections that times out are canceled and not returned to the idle pool, see test verifying it
```diff
diff --git a/staging/src/k8s.io/client-go/rest/connection_test.go b/staging/src/k8s.io/client-go/rest/connection_test.go
index c5027a0042b..7754c984f6b 100644
--- a/staging/src/k8s.io/client-go/rest/connection_test.go
+++ b/staging/src/k8s.io/client-go/rest/connection_test.go
@@ -23,6 +23,7 @@ import (
        "net"
        "net/http"
        "net/http/httptest"
+       "net/http/httptrace"
        "net/url"
        "os"
        "strconv"
@@ -43,6 +44,7 @@ type tcpLB struct {
 }
 
 func (lb *tcpLB) handleConnection(in net.Conn, stopCh chan struct{}) {
+       lb.t.Logf("Received connection from %v", in.RemoteAddr())
        out, err := net.Dial("tcp", lb.serverURL)
        if err != nil {
                lb.t.Log(err)
@@ -237,3 +239,113 @@ func TestReconnectBrokenTCP_HTTP1(t *testing.T) {
                t.Fatalf("expected %d dials, got %d", 2, dials)
        }
 }
+
+// 1. connect to https server with http1.1 using a TCP proxy making the connection to timeout
+// 2. the connection has keepalive enabled so it will be reused
+// 3. close the in-flight connection to force creating a new connection and avoid it goes to the idle pool once times out
+// 4. count that there are 2 connection on the LB but only one succeeds
+func TestReconnectBrokenTCPAfterTimeout_HTTP1(t *testing.T) {
+       ts := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+               // hold the connection
+               time.Sleep(2 * time.Second)
+               fmt.Fprintf(w, "Hello, %s", r.Proto)
+       }))
+       ts.EnableHTTP2 = false
+       ts.StartTLS()
+       defer ts.Close()
+
+       u, err := url.Parse(ts.URL)
+       if err != nil {
+               t.Fatalf("failed to parse URL from %q: %v", ts.URL, err)
+       }
+       lb := newLB(t, u.Host)
+       defer lb.ln.Close()
+       stopCh := make(chan struct{})
+       go lb.serve(stopCh)
+       transport, ok := ts.Client().Transport.(*http.Transport)
+       if !ok {
+               t.Fatalf("failed to assert *http.Transport")
+       }
+       config := &Config{
+               Host:      "https://" + lb.ln.Addr().String(),
+               Transport: utilnet.SetTransportDefaults(transport),
+               Timeout:   3 * time.Second,
+               // These fields are required to create a REST client.
+               ContentConfig: ContentConfig{
+                       GroupVersion:         &schema.GroupVersion{},
+                       NegotiatedSerializer: &serializer.CodecFactory{},
+               },
+       }
+       config.TLSClientConfig.NextProtos = []string{"http/1.1"}
+       client, err := RESTClientFor(config)
+       if err != nil {
+               t.Fatalf("failed to create REST client: %v", err)
+       }
+
+       reqErrCh := make(chan error, 1)
+       defer close(reqErrCh)
+       trace := &httptrace.ClientTrace{
+               GotConn: func(connInfo httptrace.GotConnInfo) {
+                       t.Logf("Got Conn Reused %v WasIdle %v IdleTime %v", connInfo.Reused, connInfo.WasIdle, connInfo.IdleTime)
+               },
+               ConnectDone: func(network, addr string, err error) {
+                       t.Logf("Connection done: %s %s", network, addr)
+               },
+               WroteRequest: func(reqInfo httptrace.WroteRequestInfo) {
+                       reqErrCh <- reqInfo.Err
+               },
+       }
+
+       // Create a connection and make it timeout closing the LB connection
+       // after it connects to the server
+       ctx := httptrace.WithClientTrace(context.Background(), trace)
+       errCh := make(chan error)
+       go func() {
+               _, err := client.Get().AbsPath("/").DoRaw(ctx)
+               errCh <- err
+       }()
+
+       select {
+       case err := <-reqErrCh:
+               if err != nil {
+                       t.Fatalf("Connection failed: %v", err)
+               }
+               // Deliberately let the LB stop proxying traffic for the current
+               // connection. This mimics a broken TCP connection that's not properly
+               // closed.
+               close(stopCh)
+
+               stopCh = make(chan struct{})
+               go lb.serve(stopCh)
+
+       case <-time.After(10 * time.Second):
+               t.Fatalf("Test timed out waiting for first request to fail")
+       }
+
+       // Wait for the connection to time out
+       select {
+       case err := <-errCh:
+               if err == nil {
+                       t.Fatalf("Connection succeeded but was expected to timeout")
+               }
+               t.Logf("Connection failed with error: %v", err)
+       case <-time.After(10 * time.Second):
+               t.Fatalf("Test timed out waiting for first request to fail")
+       }
+
+       // If the client didn't close the idle connections, the broken connection
+       // would still be in the connection pool, the following request would
+       // then reuse the broken connection instead of creating a new one, and
+       // thus would fail.
+       data, err := client.Get().AbsPath("/").DoRaw(ctx)
+       if err != nil {
+               t.Fatalf("unexpected err: %v", err)
+       }
+       if string(data) != "Hello, HTTP/1.1" {
+               t.Fatalf("unexpected response: %s", data)
+       }
+       dials := atomic.LoadInt32(&lb.dials)
+       if dials != 2 {
+               t.Fatalf("expected %d dials, got %d", 2, dials)
+       }
+}
```

Reply:
I'm going to add the test so we can avoid regressions on golang , the test seems stable.

```
stress ./rest.test -test.run "TestReconnect*"
5s: 0 runs so far, 0 failures
10s: 24 runs so far, 0 failures
15s: 24 runs so far, 0 failures
...
1m15s: 216 runs so far, 0 failures
1m20s: 216 runs so far, 0 failures
1m25s: 240 runs so far, 0 failures
1m30s: 264 runs so far, 0 failures
```
They are also really useful for troubleshooting these kind of problems